crates/sim2h/src/lib.rs

```diff
@@ -62,35 +95,114 @@ pub struct Sim2h {
-            stream_manager,
+            pool,
+            wss_recv,
+            msg_send,
+            msg_recv,
+            open_connections: HashMap::new(),
             num_ticks: 0,
             rrdht_arc_radius_recalc: std::time::Instant::now(),
             missing_aspects_resync: std::time::Instant::now(),
         };
 
-        debug!("Trying to bind to {}...", bind_spec);
-        let url: url::Url = bind_spec.clone().into();
-        sim2h.bound_uri = Some(
-            sim2h
-                .stream_manager
-                .bind(&url)
-                .unwrap_or_else(|e| panic!("Error binding to url {}: {:?}", bind_spec, e))
-                .into(),
-        );
+        sim2h.priv_bind_listening_socket(url::Url::from(bind_spec).into(), wss_send);
 
         sim2h
     }
 
+    /// bind a listening socket, and set up the polling job to accept connections
+    fn priv_bind_listening_socket(
+        &mut self,
+        url: Url2,
+        wss_send: crossbeam_channel::Sender<TcpWss>,
+    ) {
+        let config = TcpBindConfig::default();
+        let config = TlsBindConfig::new(config).dev_certificate();
+        let config = WssBindConfig::new(config);
+        let listen: TcpWssServer = InStreamListenerWss::bind(&url, config).unwrap();
+        self.bound_uri = Some(url::Url::from(listen.binding()).into());
+        self.pool
+            .push_job(Box::new(Arc::new(Mutex::new(ListenJob::new(
+                listen, wss_send,
+            )))));
+    }
+
+    /// if our listening socket has accepted any new connections, set them up
+    fn priv_check_incoming_connections(&mut self) {
+        if let Ok(wss) = self.wss_recv.try_recv() {
+            let url: Lib3hUri = url::Url::from(wss.remote_url()).into();
+            let job = Arc::new(Mutex::new(ConnectionJob::new(wss, self.msg_send.clone())));
+            if let Err(error) = self.handle_incoming_connect(url.clone()) {
+                error!("Error handling incoming connection: {:?}", error);
+                return;
+            }
+            self.open_connections.insert(url, job.clone());
+            self.pool.push_job(Box::new(job));
+        }
+    }
+
+    /// we received some kind of error related to a stream/socket
+    /// print some debugging and disconnect it
+    fn priv_drop_connection_for_error(&mut self, uri: Lib3hUri, error: Sim2hError) {
+        error!(
+            "Transport error occurred on connection to {}: {:?}",
+            uri, error,
+        );
+        info!("Dropping connection to {} because of error", uri);
+        self.disconnect(&uri);
+    }
+
+    /// if our connections sent us any data, process it
+    fn priv_check_incoming_messages(&mut self) {
+        if let Ok((url, msg)) = self.msg_recv.try_recv() {
+            let url: Lib3hUri = url::Url::from(url).into();
+            match msg {
+                Ok(frame) => match frame {
+                    WsFrame::Text(s) => self.priv_drop_connection_for_error(
```

Comment:
Would this custom protocol be replaced by something like capnproto in the future?

Reply:
No, this is the websocket protocol. It's 100% analog to the tungstenite enum - just abstracts it so we can swap out our underlying websocket lib in the future with less trouble if we need to for some reason.